Dining/models.py

```diff
@@ -204,54 +162,55 @@ def external_dining_entries(self):
+            if self.sign_up_deadline and self.sign_up_deadline.date() > self.date:
+                raise ValidationError(
+                    {'sign_up_deadline': ["Sign up deadline can't be later than the day dinner is served"]})
 
-class DiningEntry(models.Model):
-    """
-    Represents an entry on a dining list.
 
-    Do not change dining_list, user and added_by after creation!
-    """
+class DiningEntry(models.Model):
+    """Represents an entry on a dining list"""
 
-    # Dining list value should never be changed
-    dining_list = models.ForeignKey(DiningList, on_delete=models.CASCADE, related_name='dining_entries')
-    # User value should never be changed, is responsible for the money required
-    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
+    dining_list = models.ForeignKey(DiningList, on_delete=models.PROTECT, related_name='dining_entries')
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.PROTECT)
+    created_by = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.PROTECT,
+                                   related_name='created_dining_entries')
 
     has_paid = models.BooleanField(default=False)
 
-
-    def clean(self):
-        """
-        This actually has a race condition problem which makes it possible to create more entries than max_diners, and
-        it is also possible to add yourself multiple times! This is however very hard to prevent, although we could use
-        a mutex lock for the time between validation and saving.
-        """
-        if not self.pk:
-            # Validate user is not already subscribed for the dining list
-            if self.get_internal() and self.dining_list.internal_dining_entries().filter(user=self.user).exists():
-                raise ValidationError(gettext('This user is already subscribed to the dining list'))
-
-            # (Optionally) validate if user is not already on another dining list
-            # if DiningList.objects.filter(date=self.dining_list.date, dining_entries__user=self.user)
-
-    def get_internal(self):
+    def get_subclass(self):
+        """Return an instance of the correct subclass, either DiningEntryUser or DiningEntryExternal"""
         try:
             return self.diningentryuser
         except DiningEntryUser.DoesNotExist:
-            return None
-
-    def get_external(self):
+            pass
         try:
             return self.diningentryexternal
-        except DiningEntryExternal.DoesNotExist:
-            return None
+        except DiningEntryUser.DoesNotExist:
+            pass
+        raise RuntimeError("Invalid DiningEntry")
 
-    def name(self):
-        external = self.get_external()
-        if external:
-            return external.name
-        else:
-            return self.user
+    def get_name(self):
+        """Return name of diner"""
+        return self.user.get_full_name()
+
+    def is_internal(self):
+        return True
```

Comment:
What's the problem here?

Reply:
To elaborate, this works by method overriding

Reply:
RIP can_add_diner method. 8/5/19 - 10/5/19

Reply:
I don't think it will overwrite the method. When getting objects from the DiningEntry table (which Django does when going over the relation from dining list or whatever) python recognises the object as a DiningEntry object, not as a DiningEntryUser object, thus returning True on is_internal will also return True when the object is a parent of DiningEntryExternal instead.

Reply:
The get_subclass method is used for that, to ensure that the class is always correct. It's not ideal since you should always call that method first when working with the general DiningEntry, but after that you don't need to worry about that anymore in the rest of the code and can use the full potential of subclassing, which is also advantageous for each change that needs to be made in the future.

Reply:
I must've missed the get_subclass method. If that does what it should do, yeah, then it works.